### PR TITLE
Fixes reboot-and-wait-for strategy across all platforms

### LINKED ISSUE #951: avoid extra reboots
Currently in the develop version, under certain situations, the playbook will cause 2 reboots in the common role. 1 during installing the grsec kernel and 1 during the cron-apt task to perform system upgrades. We should either reorganize the order of these tasks or change the reboot tasks to an ansible handler to avoid the extra reboot.

### PR DESCRIPTION
**Pull request by conorsch**
## Status

Ready for review.

## Description of Changes

Fixes #951. Changes introduced in #1666 caused variability in the "reboot and wait for" strategy (RAWFS) used in several places throughout the site config. To merge #1666 and boost the automated tests, we opted for the RAWFS that worked well with AWS EC2 instances—but failed on Vagrant  local VMs. The new strategy works equally well in both contexts.

Changes proposed in this pull request:

* Notifies a single handler for reboot-and-wait-for (DRY)
* Reduces number of reboots from 3 to 1 on clean staging install

There are some non-obvious side-effects of the handler approach, namely:

  1. The handler is declared in the `common` role and therefore is only available to roles within the same play. 
  2. The only way to notify a handler is to have a task that reports changed=true; thus the dummy tasks that simply run `/bin/true` based on file existence.

Down the road I'd like to move toward leaning more heavily on upstream for the Ansible config, which would allow us to remove many of the custom roles for common hardening actions, e.g. sysctl and iptables management. In particular, we have a [well-maintained grsecurity role](https://github.com/freedomofpress/ansible-role-grsecurity) that we're _not_ using in this project yet. Once we include that work, we'll be in a much stronger position. 

## Testing

  1. Confirm CircleCI staging env passes.
  2. Confirm a fresh staging build on Vagrant VMs passes.

## Deployment

The changes presented here modify the implementation of reboots in the playbooks, but with an eye toward preserving existing functionality. That said, vigorous QA on hardware is warranted, although I suggest we include that on the general 0.4 QA testing list.

### If you made changes to the app code:

- [x] Unit and functional tests pass on the development VM
- [ ] Unit and functional tests pass on the app-staging VM

### If you made changes to the system configuration:

**Note:** Had to bump Ansible down to v2.2.1, since 2.2.2 showed errors with testinfra. If you receive the same results, say so, and we'll make sure to address in #1146.

- [x] Testinfra tests pass on the development VM
- [ ] Testinfra tests pass on the app-staging VM
- [ ] Testinfra tests pass on the mon-staging VM

### CODE REVIEW

**Comment by conorsch** [comment]
Should there be any trouble running the testinfra tests locally, please provide error output. I suspect we'll want to pin ansible==2.2.1 as part of #1146 but would appreciate more data first.

**Comment by redshiftzero** [comment]
OK tried to take this for a spin:

- [x] Unit and functional tests pass on the development VM
- [ ] Unit and functional tests pass on the app-staging VM
- [x] Testinfra tests pass on the development VM (on Ansible 2.2.1)
- [ ] Testinfra tests pass on the app-staging VM
- [ ] Testinfra tests pass on the mon-staging VM

I ran into timeouts while waiting for the servers to come back:
```
TASK [grsecurity : Wait for server to come back.] ******************************
fatal: [app-staging -> localhost]: FAILED! => {"changed": false, "elapsed": 300, "failed": true, "msg": "Timeout when waiting for localhost:22"}
fatal: [mon-staging -> localhost]: FAILED! => {"changed": false, "elapsed": 300, "failed": true, "msg": "Timeout when waiting for localhost:22"}
```

Currently rerunning development testinfra on Ansible 2.2.2 (I got errors there) and will paste the output here in a sec

**Comment by redshiftzero** [comment]
Error output from running testinfra tests on development on this branch on Ansible 2.2.2 (works fine on Ansible 2.2.1):

```
==================================================================== ERRORS ====================================================================
_____________________________________________ ERROR collecting testinfra/app-code/test_haveged.py ______________________________________________
/usr/local/lib/python2.7/site-packages/testinfra/plugin.py:136: in pytest_generate_tests
    ansible_inventory=metafunc.config.option.ansible_inventory,
/usr/local/lib/python2.7/site-packages/testinfra/host.py:125: in get_hosts
    for backend in testinfra.backend.get_backends(hosts, **kwargs):
/usr/local/lib/python2.7/site-packages/testinfra/backend/__init__.py:84: in get_backends
    for name in klass.get_hosts(host, **kw):
/usr/local/lib/python2.7/site-packages/testinfra/backend/ansible.py:82: in get_hosts
    return AnsibleRunner(kwargs.get("ansible_inventory")).get_hosts(host)
/usr/local/lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:165: in __init__
    host_list=host_list or self.cli.options.inventory,
/usr/local/lib/python2.7/site-packages/ansible/inventory/__init__.py:97: in __init__
    self.parse_inventory(host_list)
/usr/local/lib/python2.7/site-packages/ansible/inventory/__init__.py:181: in parse_inventory
    host.remove_group(ungrouped)
E   AttributeError: 'Host' object has no attribute 'remove_group'
============================================================ pytest-warning summary ============================================================
WP1 None Module already imported so can not be re-written: testinfra
============================================ 68 skipped, 1 pytest-warnings, 1 error in 1.78 seconds ============================================
Traceback (most recent call last):
  File "./testinfra/test.py", line 129, in <module>
    run_testinfra(target_host)
  File "./testinfra/test.py", line 126, in run_testinfra
    subprocess.check_call(testinfra_command)
  File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['testinfra', '-vv', '-n', 'auto', '--connection', 'ansible', '--ansible-inventory', '.vagrant/provisioners/ansible/inventory/vagrant_ansible_inventory', '--hosts', 'development', 'testinfra/app-code', 'testinfra/development']' returned non-zero exit status 1
```

File: install_files/ansible-base/roles/common/tasks/apt_upgrade.yml [hunk 1]
```diff
@@ -40,31 +40,18 @@
...
-  when: reboot_required.stat.exists
-  tags:
-    - reboot
-
-- name: Wait for server to come back.
-  local_action: wait_for
+  # Ansible doesn't support notifying handlers based on file existence, results,
+  # e.g. via registered `stat` results, only only on changed=true task results,
+  # so let's craft a task that returns changed=true if-and-only-if a file exists.
+- name: Check if reboot is required due to security updates.
```
Comment (msheiny) [inline install_files/ansible-base/roles/common/tasks/apt_upgrade.yml]: :laughing: this task is ansible hackery at it's finest. :champagne: no seriously, i like the work-around here - its sweet how adapt you've gotten at working around some of ansible's current limitations.

**Comment by redshiftzero** [comment]
OK sorry for delay on this. Wanted to try this with each configuration of Ansible
and do it a couple of times so that it worked. I get failures on all three testinfra suites on Ansible 2.2.2 with identical tracebacks. I'm provisioning staging VMs on this branch and doing a `vagrant reload`. 

# QA 

Links go to gists with error output for each point.

- [ ] Application tests on app-staging
- [ ] [Testinfra tests on app-staging VM on Ansible 2.2.1](https://gist.github.com/redshiftzero/3016ea21e5eafcc84240a6d99687e5fa#file-app-staging-ansible-2-2-1)
- [ ] [Testinfra tests on mon-staging VM on Ansible 2.2.1](https://gist.github.com/redshiftzero/3016ea21e5eafcc84240a6d99687e5fa#file-mon-staging-ansible-2-2-1)
- [ ] [Testinfra tests on app-staging VM on Ansible 2.2.2](https://gist.github.com/redshiftzero/3016ea21e5eafcc84240a6d99687e5fa#file-mon-staging-ansible-2-2-2)
- [ ] [Testinfra tests on mon-staging VM on Ansible 2.2.2](https://gist.github.com/redshiftzero/3016ea21e5eafcc84240a6d99687e5fa#file-mon-staging-ansible-2-2-2)

After doing some digging around, the Ansible error that I'm getting all over those tracebacks when using Ansible 2.2.1 appears to be one that [others are having on the Mac OS X platform when using Ansible 2](https://github.com/ansible/ansible/issues/12259). There is a mitigation suggested [here](https://docs.ansible.com/ansible/intro_installation.html#control-machine-requirements) and I've applied this and tried again and am still seeing the same errors.

**Comment by redshiftzero** [comment]
OK I was able to resolve all but one testinfra failure following what is now described in PR #1705 (adding instructions to the docs about these Ansible 2 Mac-related failures for any other Mac developers that try to provision staging and start pulling their hair out).

I'm now stuck with just this one failed test on both [mon-staging](https://gist.github.com/redshiftzero/b7dee2ae9668023e5c7fadfa6f9b3056#file-mon-staging-error) and [app-staging](https://gist.github.com/redshiftzero/b7dee2ae9668023e5c7fadfa6f9b3056#file-app-staging-error) on testinfra (and some errors in the application functional tests, but I will get there once I'm sure nothing is funky with the machines being provisioned correctly).

SSHing into the machines to dig into the network test failure a bit, it looks in the `testinfra` suite in `test_app_iptables_rules` there is an expected environment variable `$CI_SD_ENV`  that is undefined, leading to a failure (see the "No such file or directory... iptables-app-None.j2" in the tracebacks). However if I manually set:

```
    environment = "staging"
    # environment = os.environ.get("CI_SD_ENV")  # comment this out to see if it's the envvar
    iptables_file = "{}/iptables-mon-{}.j2".format(
                                      os.path.dirname(os.path.abspath(__file__)),
                                      environment)
```

then this final test now passes... 

(...will continue to dig unless I reach a resolution, am just being slightly verbose in case any of this has been seen by anyone else)

**Comment by conorsch** [comment]
Looks to me like that testinfra test can't be expected to pass in staging, since it should be using `environment = os.environ.get("CI_SD_ENV", "staging")` to provide a default value properly. Probably was missed since #1666 broke staging locally, so I expect there to be some follow-up cleanup to testinfra.

**Comment by redshiftzero** [comment]
# QA Round 2

- [ ] Application tests on app-staging
- [x] Testinfra tests on app-staging VM on Ansible 2.2.1
- [x] Testinfra tests on mon-staging VM on Ansible 2.2.1

Passing testinfra locally after 71dcb98 was added. Now on to debugging the application test failures..

**Comment by redshiftzero** [comment]
Testinfra is now reliably passing for me ✅. Took another look at the application errors, and I'm getting failures on the following application tests:

```
tests/test_source.py::TestSourceApp::test_custom_notification FAILED
tests/functional/test_admin_interface.py::AdminInterface::test_admin_interface FAILED
tests/functional/test_submit_and_retrieve_file.py::SubmitAndRetrieveFile::test_source_cancels_at_submit_page FAILED
tests/functional/test_submit_and_retrieve_file.py::SubmitAndRetrieveFile::test_submit_and_retrieve_happy_path FAILED
tests/functional/test_submit_and_retrieve_message::SubmitAndRetrieveMessage::test_submit_and_retrieve_happy_path FAILED
```

Digging into what's going on, it appears there are two underlying problems producing the functional test failures:
1) The files in `/var/www/securedrop/static/.webassets-cache` are readable only by root. If you look at `/var/log/apache2/source-error.log` after running the functional tests you should see a bunch of errors like this one:

```
IOError: [Errno 13] Permission denied: '/var/www/securedrop/static/.webassets-cache/7d262489daab354fc856a9250e4c560e'
```

2) The generated CSS is not present in the target directory: `/var/www/securedrop/static/css`

**Comment by conorsch** [comment]
> Digging into what's going on, it appears there are two underlying problems producing the functional test failures

Fairly certain those failures exist on the develop branch, and we haven't noticed them on local staging due to the reboot problems. If so, we can merge this PR to satisfy the reboot logic updates, then open separate issues to tackle the asset perms in the deb packages.

This PR needs a rebase, due to merge of #1703. Will handle that shortly.

**Comment by conorsch** [comment]
Rebased on top of latest develop. All config tests are passing for me locally, which is great news. (Had to cherry-pick ef1eed8ccf440ac0c52e8e145dddeb32a7a778d3 after rebasing, since I failed to pull it down from remote prior to rebasing.)

Can confirm the app test failures @redshiftzero described, so I recommend we punt those to another issue, since the new reboot logic is required to drill down on fixing those problems (and updating the _config_ tests where appropriate too).

**Comment by redshiftzero** [comment]
OK tested this latest version, all testinfra tests passing on staging. :+1: from me to merge this, and we can tackle the application test failures on staging in another issue

**Pull request merged**